src/xwayland/xserver.rs

```diff
@@ -0,0 +1,358 @@
+    }
+}
+
+fn xwayland_ready<WM: XWindowManager>(inner: &Rc<RefCell<Inner<WM>>>) {
+    use nix::sys::wait;
+    let mut guard = inner.borrow_mut();
+    let inner = &mut *guard;
+    // instance should never be None at this point
+    let instance = inner.instance.as_mut().unwrap();
+    let wm = &mut inner.wm;
+    // neither the pid
+    let pid = instance.child_pid.unwrap();
+
+    // find out if the launch was a success by waiting on the intermediate child
+    let mut success: bool;
+    loop {
+        match wait::waitpid(pid, None) {
+            Ok(wait::WaitStatus::Exited(_, 0)) => {
+                // XWayland was correctly started :)
+                success = true;
+                break;
+            }
+            Err(NixError::Sys(Errno::EINTR)) => {
+                // interupted, retry
+                continue;
+            }
+            _ => {
+                // something went wrong :(
+                success = false;
+                break;
+            }
+        }
+    }
+
+    if success {
+        // signal the WM
+        wm.xwayland_ready(
+            instance.wm_fd.take().unwrap(), // This is a bug if None
+            instance.wayland_client.clone(),
+        );
+
+        // setup the environemnt
+        ::std::env::set_var("DISPLAY", format!(":{}", instance.display_lock.display()));
+    }
+
+    // in all cases, cleanup
+    if let Some(s) = instance.sigusr1_handler.take() {
+        s.remove();
+    }
+}
+
+enum Void {}
+
+/// Exec xwayland with given sockets on given display
+///
+/// If this returns, that means that something failed
+fn exec_xwayland(
+    display: u32,
+    wayland_socket: UnixStream,
+    wm_socket: UnixStream,
+    listen_sockets: &[UnixStream],
+) -> NixResult<Void> {
+    // uset the CLOEXEC flag from the sockets we need to pass
+    // to xwayland
+    unset_cloexec(&wayland_socket)?;
+    unset_cloexec(&wm_socket)?;
+    for socket in listen_sockets {
+        unset_cloexec(socket)?;
+    }
+    // prepare the arguments to Xwayland
+    let mut args = vec![
+        CString::new("Xwayland").unwrap(),
+        CString::new(format!(":{}", display)).unwrap(),
+        CString::new("-rootless").unwrap(),
+        CString::new("-terminate").unwrap(),
+        CString::new("-wm").unwrap(),
+        CString::new(format!("{}", wm_socket.as_raw_fd())).unwrap(),
+    ];
+    for socket in listen_sockets {
+        args.push(CString::new("-listen").unwrap());
+        args.push(CString::new(format!("{}", socket.as_raw_fd())).unwrap());
+    }
+    // setup the environment: clear everything except PATH and XDG_RUNTIME_DIR
+    for (key, _) in env::vars_os() {
+        if key.to_str() == Some("PATH") || key.to_str() == Some("XDG_RUNTIME_DIR") {
+            continue;
+        }
+        env::remove_var(key);
+    }
+    // the WAYLAND_SOCKET var tells Xwayland where to connect as a wayland client
+    env::set_var("WAYLAND_SOCKET", format!("{}", wayland_socket.as_raw_fd()));
```

Comment:
And environment variable with the number of our file descriptor? This is really how this works?

Reply:
Well, yeah: https://github.com/wayland-project/wayland/blob/master/src/wayland-client.c#L1128